Decoupled connection roles from ConnectionManager

othercorey opened the pull request:
replaces https://github.com/cakephp/cakephp/pull/16785

After implementing the connection name solution for read/write roles, we discovered a couple limitations that would break manually created connections. This implements an original proposal but cleaned up and simplified.

## Config

Users can now specify a `read` and `write` key in their connection config. 

* If set, these values override the shared config in the top-level config and a separate Driver instance is created.
* If the configs are identical, the one driver is created and shared the same as if `read` and `write` didn't exist.
* The `driver` key is not customizable per role so users cannot accidentally create drivers from two different architectures for one connection.
* The `driver` key can still be set to a driver instance and will be used as both read and write role regardless of other config.

## Use

Users can only change the connection role for `SelectQuery` instances. All other queries default to the `write` role and cannot be changed.

You can set it using:

```php
$selectQuery->setConnectionRole(Connection::ROLE_READ)
```

## Implementation

All operations that affect write-only features default to using the write role. This includes getting table schemas, transactions, save points and foreign key flags.

Users must pass the connection role from the query to `Connection::getDriver()` if they're expecting a non-default role.

```php
$query->getConnection()->getDriver($query->getConnectionRole())
```

othercorey committed:
Decoupled connection roles from ConnectionManager

othercorey commented on src/Database/Connection.php:
```diff
@@ -1085,9 +1141,16 @@ public function __debugInfo(): array
+        }
+        if (isset($config['write'])) {
+            $config['write'] = array_intersect_key($secrets, $config['write']) + $config['write'];
+        }
```
I could not find a test case for sanitizing.

othercorey commented:
I need to add a test case for the config.

ADmad commented on src/Database/Connection.php:
```diff
@@ -139,19 +139,63 @@ class Connection implements ConnectionInterface
 
-        $driverConfig = array_diff_key($config, array_flip([
+        if (!empty($config['log'])) {
+            $this->enableQueryLogging((bool)$config['log']);
```
The query log should now include the role.

ADmad commented on src/Database/Connection.php:
```diff
@@ -60,9 +60,9 @@ class Connection implements ConnectionInterface
+     * @var array<string, \Cake\Database\DriverInterface>
      */
-    protected $_driver;
+    protected array $_drivers;
```
If we use separate properties instead of an array we can type them in 5.x and potentially avoid the asserts() we have in `createDrivers()`.

othercorey replied:
After changing everything to `getDriver()` this seems much less complicated than before.

ADmad commented on src/Database/Connection.php:
```diff
@@ -269,20 +317,25 @@ public function getDriver(): DriverInterface
-                $e
-            );
+        $connected = true;
+        foreach ([self::ROLE_READ, self::ROLE_WRITE] as $role) {
```
Don't really like connecting both drivers when potentially only 1 could be needed. `ReconnectStrategy` is the only place where `Connection::connect()` is used. So in 5.x we should explore changes to avoid connecting both the drivers.

othercorey replied:
Agreed, but I wasn't sure how complicated that would be so didn't change in this PR.

markstory commented on src/Database/Connection.php:
```diff
@@ -139,19 +139,63 @@ class Connection implements ConnectionInterface
+    {
+        $driver = $config['driver'] ?? '';
+        if (!is_string($driver)) {
+            assert($driver instanceof DriverInterface);
```
```suggestion
            assert($driver instanceof DriverInterface, 'Drivers must implement \Cake\Database\DriverInterface');
```

markstory commented on src/Database/Connection.php:
```diff
@@ -254,11 +299,14 @@ public function getDisconnectRetry(): CommandRetry
      * @return \Cake\Database\DriverInterface
      */
-    public function getDriver(): DriverInterface
+    public function getDriver(string $role = self::ROLE_WRITE): DriverInterface
```
This could end up breaking userland implementations. We currently have `getDriver()` annotated only, but there could be a userland implementation of `ConnectionInterface`

othercorey replied:
You said last PR that adding parameters with a default to a public function was allowed in the cake minor release policy.

markstory replied:
Derp. You're right as the method isn't on the interface.

markstory commented on src/Database/Connection.php:
```diff
@@ -313,8 +367,10 @@ public function isConnected(): bool
-            $statement = $this->_driver->prepare($query);
+        $role = $query instanceof Query ? $query->getConnectionRole() : self::ROLE_WRITE;
+
+        return $this->getDisconnectRetry()->run(function () use ($query, $role) {
```
If `getDisconnectRetry()` took a role name we could skip having to connect both connections.

markstory commented on src/Database/Query.php:
```diff
@@ -231,48 +237,13 @@ public function getConnection(): Connection
-     *
-     * @return $this
-     */
-    public function useReadRole()
```
I get that replica usage is being confined to `SelectQuery` which is great. However, these methods are easier to use than `setConnectionRole()` is as `setConnectionRole()` requires importing Connection to get to the constants.

markstory commented on src/Database/Query/SelectQuery.php:
```diff
@@ -89,4 +90,18 @@ public function set($key, $value = null, $types = [])
+     * @param string $role Connection role ('read' or 'write')
+     * @return $this
+     */
+    public function setConnectionRole(string $role)
```
We could move `useReadRole()` and `useWriteRole()` here.

othercorey replied:
We can. I was trying to keep "connection" in the name to make it clear. I thought `useReadConnectionRole()` was too much, but no reason we can't add it.

markstory replied:
What about `useReadConnection()`?

othercorey replied:
We could but since this is now self-contained within a single connection, I thought that would be misleading.

othercorey replied:
I'll add whatever you think is best here. This is just the thought process I went through.

markstory replied:
Lets go with `useReadRole()` and `useWriteRole()` then. It saves userland code having to import Connection or ConnectionInterface to get the constants.

othercorey commented:
* Split Connection::$_drivers into $readDriver and $writeDriver
* Passed _role config to Driver constructor and added `getRole()` to Driver
* Added "role=" to QueryLogger
* Added useReadRole() and useWriteRole() to SelectQuery

othercorey commented on tests/TestCase/Database/ConnectionTest.php:
```diff
@@ -196,6 +194,16 @@ public function testDriverOptionClassNameSupport(): void
+        $this->assertNotSame($connection->getDriver(Connection::ROLE_READ), $connection->getDriver(Connection::ROLE_WRITE));
+        $this->assertSame(Connection::ROLE_READ, $connection->getDriver(Connection::ROLE_READ)->getRole());
+        $this->assertSame(Connection::ROLE_WRITE, $connection->getDriver(Connection::ROLE_WRITE)->getRole());
+    }
```
Added test for split read/write configs.

othercorey commented on src/Database/DriverInterface.php:
```diff
@@ -28,6 +28,7 @@
  * @method bool supports(string $feature) Checks whether a feature is supported by the driver.
  * @method bool inTransaction() Returns whether a transaction is active.
  * @method array config() Get the configuration data used to create the driver.
+ * @method string getRole() Returns the connection role this driver prforms.
```
We expect drivers to extend Driver class even with this interface.

othercorey commented on src/Database/Driver.php:
```diff
@@ -523,6 +523,16 @@ public function getConnectRetries(): int
+    public function getRole(): string
+    {
+        return $this->_config['_role'] ?? Connection::ROLE_WRITE;
+    }
```
Should this be getConnectionRole to match Query?

markstory replied:
I don't think it needs to match. Query's benefit from the additional scoping as `getRole()` by itself could be related to application logic roles or something else. I don't think we have that ambiguity here.

markstory reviewed (approved):
Looks good to me. Nice work on simplifying this :clap:
